channeldb/control_tower.go

```diff
@@ -0,0 +1,487 @@
+		// re-attempting.
+		return bucket.Delete(paymentFailInfoKey)
+	})
+	if err != nil {
+		return nil
+	}
+
+	return updateErr
+}
+
+// RegisterAttempt atomically records the provided PaymentAttemptInfo to the
+// DB.
+func (p *paymentControl) RegisterAttempt(paymentHash lntypes.Hash,
+	attempt *PaymentAttemptInfo) error {
+
+	// Serialize the information before opening the db transaction.
+	var a bytes.Buffer
+	if err := serializePaymentAttemptInfo(&a, attempt); err != nil {
+		return err
+	}
+	attemptBytes := a.Bytes()
+
+	var updateErr error
+	err := p.db.Batch(func(tx *bbolt.Tx) error {
+		// Reset the update error, to avoid carrying over an error
+		// from a previous execution of the batched db transaction.
+		updateErr = nil
+
+		bucket, err := fetchPaymentBucket(tx, paymentHash)
+		if err == ErrPaymentNotInitiated {
+			updateErr = ErrPaymentNotInitiated
+			return nil
+		} else if err != nil {
+			return err
+		}
+
+		// Add the payment attempt to the payments bucket.
+		return bucket.Put(paymentAttemptInfoKey, attemptBytes)
+	})
+	if err != nil {
+		return err
+	}
+
+	return updateErr
+}
+
+// Success transitions a payment into the Succeeded state. After invoking this
+// method, InitPayment should always return an error to prevent us from making
+// duplicate payments to the same payment hash. The provided preimage is
+// atomically saved to the DB for record keeping.
+func (p *paymentControl) Success(paymentHash lntypes.Hash,
+	preimage lntypes.Preimage) error {
+
+	var updateErr error
+	err := p.db.Batch(func(tx *bbolt.Tx) error {
+		// Reset the update error, to avoid carrying over an error
+		// from a previous execution of the batched db transaction.
+		updateErr = nil
+
+		bucket, err := fetchPaymentBucket(tx, paymentHash)
+		if err == ErrPaymentNotInitiated {
+			updateErr = ErrPaymentNotInitiated
+			return nil
+		} else if err != nil {
+			return err
+		}
+
+		// Get the existing status, if any.
+		paymentStatus := fetchPaymentStatus(bucket)
+
+		switch {
+
+		// Our records show the payment as unknown, meaning it never
+		// should have left the switch.
+		case paymentStatus == StatusUnknown:
+			updateErr = ErrPaymentNotInitiated
+			return nil
+
+		// The payment was already failed, and we are now reporting
+		// that it has succeeded. Return an error to not camouflague
+		// this bug.
+		case paymentStatus == StatusFailed:
+			updateErr = ErrPaymentAlreadyFailed
+			return nil
+
+		// A successful response was received for an InFlight payment,
+		// mark it as succeeded to prevent sending to this payment hash
+		// again.
+		case paymentStatus == StatusInFlight:
+
+		// The payment was succeeded previously, alert the caller that
+		// this may be a duplicate call.
+		case paymentStatus == StatusSucceeded:
+			updateErr = ErrPaymentAlreadySucceeded
+			return nil
+
+		default:
+			updateErr = ErrUnknownPaymentStatus
+			return nil
+		}
+
+		// We'll delete any failed info to.
+		err = bucket.Delete(paymentFailInfoKey)
```

Comment:
This should have been deleted by the new `InitPayment` call already?

Reply:
That's right! 👍  Fixed: https://github.com/lightningnetwork/lnd/pull/2761/commits/d987c982de97fe5d653c61ffb6cd05fc2c583c03